### DISCUSSION TITLE
What are the limitations of ServiceLB compared to MetalLB/kube-vip?

### DISCUSSION THREAD

**Discussion by barrettc** [k3s-io/k3s#9255] [Q&A] [answered]
I'm in the process of building a bare-metal k3s cluster and I'm trying to understand the differences around when I would need to use something like MetalLB instead of the built-in ServiceLB. For use case context, my cluster will be primarily receiving sensor readings via MQTT via [VerneMQ](https://vernemq.com).

Most searches I've found are people using MetalLB and even the k3s documentation [has the following note](https://docs.k3s.io/datastore/cluster-loadbalancer):
```
For exposing LoadBalancer Services, external load-balancers can be used alongside or instead of ServiceLB, but in most cases, replacement load-balancer controllers such as MetalLB or Kube-VIP are a better choice.
```
The majority of hits on ServiceLB are inquiries on how to disable it.

I'm just having trouble finding the exact reasons **why** MetalLB is recommended. What are the limitations with ServiceLB?

**Comment by brandond** [comment, marked as answer]
ServiceLB does not give you a real load-balanced VIP. As described in the documentation, it just exposes the service on node addresses, and set those node addresses as the loadbalancer addresses.

If you want a REAL load-balanced address, you need something else.
